### PR TITLE
Add support for HPO early stopping

### PR DESCRIPTION
**Pull request by salmankhurshid1**
*Issue #, if available:*

*Description of changes:*

- Adds support for external early stopping techniques, such as HPO:

    -  Captures `SIGTERM` call received for early termination
    -  Saves model on disk after each iteration using `save_intermediate_model` callback
    -  Cleans up model directory when `SIGTERM` is received and captured
    -  Added flag to enable early stopping support

Tested with `tox`, integration tests and new functional tests for single and multiple instances(CR-28179853).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 1]
```diff
@@ -28,10 +31,51 @@
...
+
+                    try:
+                        shutil.move(current_file_path, new_file_path)
+                        logging.debug("Stored intermediate trained model at {}".format(new_file_path))
+                    except shutil.Error:
+                        logging.debug("Failed to move intermediate model from {} to {}"
+                                      .format(current_file_path, new_file_path))
+                        pass
+
+        _terminate()
```
Comment (rizwangilani) [inline src/sagemaker_xgboost_container/algorithm_mode/train.py:74]: I see a bunch of nested ifs here. Is it possible to simplify this? Perhaps make it more modular by moving some of the logic to its own function?

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 2]
```diff
@@ -178,10 +222,16 @@ def train_job(train_cfg, train_dmatrix, val_dmatrix, model_dir, checkpoint_dir,
         logging.info("Resuming from iteration %s", iteration)
 
     callbacks = []
+    force_checkpoint_dir = None
     callbacks.append(checkpointing.print_checkpointed_evaluation(start_iteration=iteration))
     if checkpoint_dir:
         save_checkpoint = checkpointing.save_checkpoint(checkpoint_dir, start_iteration=iteration)
-        callbacks.append(save_checkpoint)
+    else:
+        force_checkpoint_dir = FORCE_CHECKPOINT_DIR
```
Comment (rizwangilani) [inline src/sagemaker_xgboost_container/algorithm_mode/train.py:230]: why are we reassigning the value here?

File: src/sagemaker_xgboost_container/algorithm_mode/train.py (hunk 1 above)
Comment (salmankhurshid1) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py]: Changed in new commit by creating function in train_utils.

File: src/sagemaker_xgboost_container/algorithm_mode/train.py (hunk 2 above)
Comment (salmankhurshid1) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py]: Changed this design to have early stopping support during regular checkpointing as well.

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 3]
```diff
@@ -28,10 +29,31 @@
...
+def add_sigterm_handler(model_dir, is_master):
+    """Stop training and cleanup model directory when SIGTERM is received.
+
+    Model directory is only cleaned if is_master is True. Otherwise program terminates.
+
+    :param model_dir: Directory where model is saved
+    :param is_master: True if single node training, or the current node is the master node in distributed training
+    """
+    def _terminate():
+        os._exit(0)
```
[the remark this answers is not in the document]
Comment (salmankhurshid1) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py:46]: os._exit(0) is used since it is an Async-signal safe function, while sys.exit(0) is not. Unsafe functions can sometimes not work during interruption of other unsafe calls such as system calls.

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 4]
```diff
@@ -28,10 +29,31 @@
...
+
+    def _cleanup_files(signo, frame):
+        if is_master:
+            train_utils.cleanup_dir(model_dir, MODEL_NAME)
```
[the remark this answers is not in the document]
Comment (salmankhurshid1) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py:50]: Cleanup is required because the program can be terminated during `save_intermediate_model` callback is running, which can leave `tmp` files behind in the directory. 
For the second part, currently our model loading side assumes there is only one file in the `/opt/ml/model` directory.

File: src/sagemaker_xgboost_container/checkpointing.py [hunk 5]
```diff
@@ -322,3 +322,65 @@ def callback(self, env):
...
+    def format_path(self):
+        """Return a file path to intermediate model"""
+        intermediate_path = os.path.join(self.intermediate_model_dir, self.model_name)
+        return intermediate_path
+
+    def _save_intermediate_model(self, model):
+        """Save intermediate model to intermediate model directory"""
+        with tempfile.NamedTemporaryFile(
+                dir=self.intermediate_model_dir, delete=False) as tf:
+            model.save_model(tf.name)
```
[the remark this answers is not in the document]
Comment (salmankhurshid1) [reply src/sagemaker_xgboost_container/checkpointing.py:374]: +1. There is indeed a slight difference between `model.save_model` vs `pkl.dump(bst, file)` methods. I'll make this change.

File: test/unit/algorithm_mode/test_train_utils.py [hunk 6]
```diff
@@ -45,3 +48,51 @@ def test_get_eval_metrics_and_feval():
...
 
     assert ('accuracy', .5) == test_configured_eval(binary_preds, binary_dtrain)[0]
+
+
+def test_cleanup_dir():
+    def setup(file_names):
+        test_dir = tempfile.mkdtemp()
+        for file_name in file_names:
+            test_path = os.path.join(test_dir, file_name)
+            open(test_path, 'w')
```
[the remark this answers is not in the document]
Comment (salmankhurshid1) [reply test/unit/algorithm_mode/test_train_utils.py:58]: +1. I'll make these changes.

**Comment by salmankhurshid1** [comment]
1. Correct, that's one part of the motivation. Other is if there are changes to be made to `save_checkpoint` in the future, it's easier to maintain these two as separate classes with separate logic.

2. This will not guarantee that a model is saved for multiple instances and maybe even for single instances. During training with multiple instances, there are quite a few system calls(due to socket connections) that cant be interrupted, and if interrupted, raise errors and terminate the program. In such cases, we will not be able to use the signal handler to save the model to disk. So, it's important we have the latest copy available with us beforehand.

3. I'm currently working on adding a hyperparameter so that customers can make this decision on their own(with guidance from our side through documentation etc). My plan was to add this as an extra parameter in the sagemaker python sdk such as the ones for checkpoint. Do you think it's better to add it as a training hyperparameter?

4. PR was tested with unit tests for the `cleanup_dir` function and `save_intermediate_model` class. Total functionality was tested through new functional tests which send a `SIGTERM` and `SIGKILL`. Also tested through manual testing by running E2E training + inference job on live instances. This cant be tested through integration tests, since there isn't a mechanism to send `SIGTERM` and wait for program to finish with our integration tests.

File: test/unit/test_checkpointing.py [hunk 7]
```diff
@@ -172,6 +172,28 @@ def tearDown(self):
         shutil.rmtree(self.test_dir, ignore_errors=True)
 
 
+class TestSaveIntermediateModel(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (salmankhurshid1) [reply test/unit/test_checkpointing.py:175]: Yes, they were removed because I changed how we were saving models for this class(pickling vs save_model). There are noted problems with pickling mock objects for testing, which we were doing now with this change. I tried quite a few fixes for it, but wasn't able to get it to work and removed it. This is being tested in the functional tests(CR-28179853).

I kept this here to make sure we were testing the setting up of the callback at least, but since this is being tested in functional tests we can remove it from unit tests.

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 8]
```diff
@@ -82,6 +104,9 @@ def sagemaker_train(train_config, data_config, train_path, val_path, model_dir,
     :param sm_current_host:
     :param checkpoint_config:
     """
+    hpo_early_stopping = train_config.pop("hpo_early_stopping", False)
+    hpo_early_stopping = True if hpo_early_stopping == "true" else False
```
[the remark this answers is not in the document]
Comment (salmankhurshid1) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py:108]: - There were two options for the workflow. One was to remove it before validation and set it to my own variable, and the other was to validate it and set it in `train_job` function. The reason I went with the first was because this parameter is not going to be called in `xgboost.train`, therefore, I wanted to keep it separate. However, based on the comment below, if I were to validate it in Hyperparameter_validation then I can move these checks in the `train_job` function rather than `sagemaker_train`. And, this will not require an additional parameter in the `train_job` function.
- Currently, the defined use case is only for HPO. To give it a more general name, we can name it `early_stopping_support`(based on general use case) or `save_intermediate_model`(based on functionality)?

**Comment by salmankhurshid1** [comment]
Yes, the `early_stopping_type` parameter triggers the SIGTERM. We do not want to force all users who set this parameter to face large I/O costs in case users don't have a need for this feature. For this reason, we have another parameter. As an aside, the customer doesn't have to set these both parameters to enable this feature. To enable this feature, they simply have to set our current parameter(`hpo_early_stopping`).

File: test/unit/test_checkpointing.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (edwardjkim) [reply test/unit/test_checkpointing.py:175]: The problem is our deployment pipeline currently does not run container tests in our QA package. CodeBuild runs unit + container tests in this repository, and deployment pipeline only runs integration tests. If there no integration tests, container tests should be added [here](https://github.com/aws/sagemaker-xgboost-container/tree/master/test/integration/local) in this repository so we have automated tests. Also, maybe I'm missing something, but based on your comment on `HyperparameterTuner. early_stopping_type='AUTO'` sending a SIGTERM, I don't see why we can't have integration tests.

**Comment by edwardjkim** [comment]
> As an aside, the customer doesn't have to set these both parameters to enable this feature. To enable this feature, they simply have to set our current parameter(`hpo_early_stopping`).

Sorry if I wasn't clear. Let me rephrase. Let's say a customer uses Python SDK to trigger a tuning job and sets `early_stopping_type='Auto'`:
```
HyperparameterTuner(
    xgboost_estimator,
    early_stopping_type="Auto"
)
```
Is the container going to save intermediate models if `early_stopping_type='Auto`? Do the customers also have to set `hpo_early_stopping="true"` in order to have model files for early-stopped jobs?

**Comment by salmankhurshid1** [comment]
> Is the container going to save intermediate models if `early_stopping_type='Auto`? Do the customers also have to set `hpo_early_stopping="true"` in order to have model files for early-stopped jobs?

Yes, the customer will also have to set `hpo_early_stopping="true"` even if `early_stopping_type=Auto` is set. The reason for that is if a user wants early stopping of HPO jobs through `early_stopping_type=Auto`, there might be cases(high I/O costs because of large models, no need of intermediate models, etc.) where users don't want to use our early stopping support(saving intermediate model) feature and we don't want to force it on them on every job.

**Comment by salmankhurshid1** [comment]
- Sure, I'll follow that as a separate action item.

- Those two files can be reused; my only concern regarding that would be if they're modified/removed for any reason then that would cause early stopping tests to fail as well. If we don't see that happening, then I can change the path in early stopping tests, and reuse those simply.

File: src/sagemaker_xgboost_container/algorithm_mode/train.py [hunk 9]
```diff
@@ -152,6 +174,9 @@ def train_job(train_cfg, train_dmatrix, val_dmatrix, model_dir, checkpoint_dir,
     :param model_dir: Directory where model will be saved
     :param is_master: True if single node training, or the current node is the master node in distributed training.
     """
+    # Parse arguments for intermediate model callback
+    early_stopping_support = train_cfg.pop('early_stopping_support', False)
```
[the remark this answers is not in the document]
Comment (salmankhurshid1) [reply src/sagemaker_xgboost_container/algorithm_mode/train.py:178]: +1. Let's make it consistent.

**Comment by salmankhurshid1** [comment]
Parameter behavior:
- Name set to "save_model_on_termination" based on functionality.
- Default behavior is set to disabled("false").

**Comment by edwardjkim** [comment]
Merging to have this included in the upcoming deployment.
